Modules/_blake2/blake2s_impl.c

```diff
@@ -272,24 +276,23 @@ Update this hash object's state with the provided bytes-like object.
 
 static PyObject *
 _blake2_blake2s_update(BLAKE2sObject *self, PyObject *data)
-/*[clinic end generated code: output=757dc087fec37815 input=97500db2f9de4aaa]*/
+/*[clinic end generated code: output=010dfcbe22654359 input=ffc4aa6a6a225d31]*/
 {
     Py_buffer buf;
 
     GET_BUFFER_VIEW_OR_ERROUT(data, &buf);
 
-    if (self->lock == NULL && buf.len >= HASHLIB_GIL_MINSIZE)
-        self->lock = PyThread_allocate_lock();
-
-    if (self->lock != NULL) {
-       Py_BEGIN_ALLOW_THREADS
-       PyThread_acquire_lock(self->lock, 1);
-       blake2s_update(&self->state, buf.buf, buf.len);
-       PyThread_release_lock(self->lock);
-       Py_END_ALLOW_THREADS
+    self->use_mutex = true;
```

Comment:
This pattern should instead be:

```c
if (!self->use_mutex && buf.len >= HASHLIB_GIL_MINSIZE) {
    self->use_mutex = true;
}
if (self->use_mutex) {
    Py_BEGIN_ALLOW_THREADS
    PyMutex_Lock(&self->mutex);
    blake2s_update(&self->state, buf.buf, buf.len);
    PyMutex_Unlock(&self->mutex);
    Py_END_ALLOW_THREADS
}
else {
    blake2s_update(&self->state, buf.buf, buf.len);
}
```

This keeps the default build behavior the same as before: the first time we see a big buffer, we use a mutex from then on. (And always use a mutex in nogil builds).